site/content/entry/yuzu-bcr/index.md

```diff
@@ -0,0 +1,126 @@
+
+{{< message "Profiling" >}}
+In software engineering, profiling ("program profiling", "software profiling") is a form of dynamic program analysis that measures, for example, the space (memory) or time complexity of a program, 
+the usage of particular instructions, or the frequency and duration of function calls. Most commonly, profiling information serves to aid program optimization.
+{{< /message >}}
+
+The problem lay in the fact that games aren't exactly streaming data all the time.
+So using immediate uploads (on OpenGL) and faster caching yielded much better performance than having a stream buffer and caching large resources, at least for Nvidia.
+Upon further testing, we found that this turned out to be false for non-Nvidia drivers on OpenGL (AMD, Intel, and Mesa) and hence had to add a stream buffer for small uploads in these drivers.
+
+## What's changed now?
+
+The technical design goals for the Buffer Cache Rewrite were the same as our Texture Cache Rewrite.
+
+- Cleaner code: No more virtual function calls or shared pointers, meaning easier maintenence in the future.
+- Improved efficiency and improved performance.
+
+Resolving which buffer existed in which memory region was a very expensive operation in our old buffer cache implementation.
+This is why the stream buffer existed — to make it faster.
+
+The new Buffer Cache has vastly improved tracking for the various buffers it caches.
+In the new implementation, when buffers are created in the memory, they are forcibly aligned to 4K [pages](https://en.wikipedia.org/wiki/Page_(computer_memory)) (4096 bytes - starting at zero).
+And to efficiently know what buffer exists on what address, the cache uses a flat array 32 MiB wide to translate from the current CPU page where the buffer exists to what buffer resides in it.
+`e.g., if the address is 4096 or 7000, that is page 1 & if it is 8192, that is page 2.`
+Thus, the new Buffer Cache can track what pages of a buffer have been modified on a page basis instead of being a binary state.
+
+Imagine if a buffer has a size of 524288 bytes and a game modifies only 1 byte of the buffer.
+Since buffers are now aligned to 4096 bytes as mentioned earlier, only those 4096 bytes are uploaded to the GPU.
+The same thing happens when the GPU tries to update the cache with data modified by the CPU.
+
+This tracking is done by making use of bit arrays in the buffers.
+Each value represents the state of the page - 1 being modified, 0 being clear.
+Keeping things in a bit array allows us to use efficient bit operations like `std::countr_zero` and `std::countr_one` (C++20).
+This results in fewer instructions yielding the same results (much faster).
+
+## All right, let's talk performance gains!
+
+The main focus of this work is in improving performance, but some graphical corrections also make its way thanks to the rewrite.
```

Comment:
Change `The main focus of this work is in improving performance, but some graphical corrections also make its way thanks to the rewrite.`
to 
`The main focus of this work is to improve performance, but some graphical improvements also resulted from this rewrite.`